# Tanner Bjorgan

## Research question/interests
I would like to see what languages are used the most/in high demand in this industry or what languages are needed to get into more niche careers.

In [1]:
import pandas as pd
import numpy as np

## **Milestone 2**

In [29]:
df2021 = pd.read_csv('../data/raw/survey_results_public_2021.csv')

In [30]:
df2021.head()

ResponseId                                         MainBranch  \
0           1                     I am a developer by profession   
1           2             I am a student who is learning to code   
2           3  I am not primarily a developer, but I write co...   
3           4                     I am a developer by profession   
4           5                     I am a developer by profession   

                                          Employment  \
0  Independent contractor, freelancer, or self-em...   
1                                 Student, full-time   
2                                 Student, full-time   
3                                 Employed full-time   
4  Independent contractor, freelancer, or self-em...   

                                             Country US_State UK_Country  \
0                                           Slovakia      NaN        NaN   
1                                        Netherlands      NaN        NaN   
2                                 Russian Federation      NaN        NaN   
3                                            Austria      NaN        NaN   
4  United Kingdom of Great Britain and Northern I...      NaN    England   

                                             EdLevel     Age1stCode  \
0  Secondary school (e.g. American high school, G...  18 - 24 years   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
2       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
3    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  11 - 17 years   
4    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   5 - 10 years   

                                           LearnCode YearsCode  ...  \
0  Coding Bootcamp;Other online resources (ex: vi...       NaN  ...   
1  Other online resources (ex: videos, blogs, etc...         7  ...   
2  Other online resources (ex: videos, blogs, etc...       NaN  ...   
3                                                NaN       NaN  ...   
4                            Friend or family member        17  ...   

               Age Gender Trans                Sexuality  \
0  25-34 years old    Man    No  Straight / Heterosexual   
1  18-24 years old    Man    No  Straight / Heterosexual   
2  18-24 years old    Man    No        Prefer not to say   
3  35-44 years old    Man    No  Straight / Heterosexual   
4  25-34 years old    Man    No                      NaN   

                      Ethnicity                Accessibility  \
0  White or of European descent            None of the above   
1  White or of European descent            None of the above   
2             Prefer not to say            None of the above   
3  White or of European descent  I am deaf / hard of hearing   
4  White or of European descent            None of the above   

        MentalHealth           SurveyLength                  SurveyEase  \
0  None of the above  Appropriate in length                        Easy   
1  None of the above  Appropriate in length                        Easy   
2  None of the above  Appropriate in length                        Easy   
3                NaN  Appropriate in length  Neither easy nor difficult   
4                NaN  Appropriate in length                        Easy   

  ConvertedCompYearly  
0             62268.0  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  

[5 rows x 48 columns]

In [4]:
df2021.tail()

ResponseId                      MainBranch  \
83434       83435  I am a developer by profession   
83435       83436  I am a developer by profession   
83436       83437  I am a developer by profession   
83437       83438  I am a developer by profession   
83438       83439  I am a developer by profession   

                                              Employment  \
83434                                 Employed full-time   
83435  Independent contractor, freelancer, or self-em...   
83436                                 Employed full-time   
83437                                 Employed full-time   
83438                                 Employed full-time   

                        Country    US_State UK_Country  \
83434  United States of America       Texas        NaN   
83435                     Benin         NaN        NaN   
83436  United States of America  New Jersey        NaN   
83437                    Canada         NaN        NaN   
83438                    Brazil         NaN        NaN   

                                                 EdLevel     Age1stCode  \
83434       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
83435       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
83436  Secondary school (e.g. American high school, G...  11 - 17 years   
83437       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  11 - 17 years   
83438                 Professional degree (JD, MD, etc.)  11 - 17 years   

                                               LearnCode YearsCode  ...  \
83434  Other online resources (ex: videos, blogs, etc...         6  ...   
83435  Other online resources (ex: videos, blogs, etc...         4  ...   
83436                                             School        10  ...   
83437  Online Courses or Certification;Books / Physic...         5  ...   
83438                                             School        14  ...   

                   Age Gender Trans                Sexuality  \
83434  25-34 years old    Man    No  Straight / Heterosexual   
83435  18-24 years old    Man    No  Straight / Heterosexual   
83436  25-34 years old    Man    No                      NaN   
83437  25-34 years old    Man    No  Straight / Heterosexual   
83438  18-24 years old    Man    No  Straight / Heterosexual   

                          Ethnicity      Accessibility  \
83434  White or of European descent  None of the above   
83435   Black or of African descent  None of the above   
83436  White or of European descent  None of the above   
83437  White or of European descent  None of the above   
83438        Hispanic or Latino/a/x  None of the above   

                                            MentalHealth  \
83434  I have a concentration and/or memory disorder ...   
83435                                  None of the above   
83436                                  None of the above   
83437  I have a mood or emotional disorder (e.g. depr...   
83438                                  None of the above   

                SurveyLength                  SurveyEase ConvertedCompYearly  
83434  Appropriate in length                        Easy            160500.0  
83435  Appropriate in length                        Easy              3960.0  
83436  Appropriate in length  Neither easy nor difficult             90000.0  
83437  Appropriate in length  Neither easy nor difficult            816816.0  
83438  Appropriate in length                        Easy             21168.0  

[5 rows x 48 columns]

## **Milestone 3**
### **Task1** : EDA

In [2]:
import seaborn as sns
import matplotlib.pylab as plt
import project_functions3 as p3
from collections import Counter

Displaying the unprossed data from the dataset to better understand the data and to know what needs to be cleaned.

In [3]:
dfU2021 = p3.loadU('../data/raw/survey_results_public_2021.csv')
dfU2021.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83439 entries, 0 to 83438
Data columns (total 48 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   ResponseId                    83439 non-null  int64  
 1   MainBranch                    83439 non-null  object 
 2   Employment                    83323 non-null  object 
 3   Country                       83439 non-null  object 
 4   US_State                      14920 non-null  object 
 5   UK_Country                    4418 non-null   object 
 6   EdLevel                       83126 non-null  object 
 7   Age1stCode                    83243 non-null  object 
 8   LearnCode                     82963 non-null  object 
 9   YearsCode                     81641 non-null  object 
 10  YearsCodePro                  61216 non-null  object 
 11  DevType                       66484 non-null  object 
 12  OrgSize                       60726 non-null  object 
 13  C

Found 83439 entries in 9 columns and shows null values in each column

In [73]:
print(f"List of the columns found: {dfU2021.columns}")

List of the columns found: Index(['ResponseId', 'MainBranch', 'Employment', 'Country', 'US_State',
       'UK_Country', 'EdLevel', 'Age1stCode', 'LearnCode', 'YearsCode',
       'YearsCodePro', 'DevType', 'OrgSize', 'Currency', 'CompTotal',
       'CompFreq', 'LanguageHaveWorkedWith', 'LanguageWantToWorkWith',
       'DatabaseHaveWorkedWith', 'DatabaseWantToWorkWith',
       'PlatformHaveWorkedWith', 'PlatformWantToWorkWith',
       'WebframeHaveWorkedWith', 'WebframeWantToWorkWith',
       'MiscTechHaveWorkedWith', 'MiscTechWantToWorkWith',
       'ToolsTechHaveWorkedWith', 'ToolsTechWantToWorkWith',
       'NEWCollabToolsHaveWorkedWith', 'NEWCollabToolsWantToWorkWith', 'OpSys',
       'NEWStuck', 'NEWSOSites', 'SOVisitFreq', 'SOAccount', 'SOPartFreq',
       'SOComm', 'NEWOtherComms', 'Age', 'Gender', 'Trans', 'Sexuality',
       'Ethnicity', 'Accessibility', 'MentalHealth', 'SurveyLength',
       'SurveyEase', 'ConvertedCompYearly'],
      dtype='object')


In [76]:
print(f"Number of unique values per column: {dfU2021.nunique(axis=0)}")

Number of unique values per column: ResponseId                      83439
MainBranch                          6
Employment                          9
Country                           181
US_State                           55
UK_Country                          4
EdLevel                             9
Age1stCode                          9
LearnCode                         458
YearsCode                          52
YearsCodePro                       52
DevType                          8626
OrgSize                            10
Currency                          146
CompTotal                        3752
CompFreq                            3
LanguageHaveWorkedWith          28690
LanguageWantToWorkWith          27545
DatabaseHaveWorkedWith           3072
DatabaseWantToWorkWith           3137
PlatformHaveWorkedWith            118
PlatformWantToWorkWith            126
WebframeHaveWorkedWith           5402
WebframeWantToWorkWith           5113
MiscTechHaveWorkedWith           1387
MiscTechWantTo

In [81]:
print(f"Unique developper types: {dfU2021.DevType.unique()} ")

Unique developper types: ['Developer, mobile' nan 'Developer, front-end' ...
 'Developer, mobile;Developer, front-end;Developer, full-stack;Developer, back-end;Database administrator;Developer, embedded applications or devices;System administrator;Educator'
 'Developer, front-end;Developer, desktop or enterprise applications;Developer, full-stack;Developer, back-end;Developer, QA or test;Developer, embedded applications or devices;Engineering manager;Educator'
 'Developer, mobile;Developer, desktop or enterprise applications;Data scientist or machine learning specialist;Developer, back-end;Engineering manager'] 


In [83]:
dfU2021['LanguageHaveWorkedWith'].describe()

count      82357
unique     28690
top       Python
freq        1337
Name: LanguageHaveWorkedWith, dtype: object

In [84]:
dfU2021.describe().round()

ResponseId     CompTotal  ConvertedCompYearly
count     83439.0  4.718300e+04              46844.0
mean      41720.0  2.119407e+69             118426.0
std       24087.0  4.603702e+71             527294.0
min           1.0  0.000000e+00                  1.0
25%       20860.0  1.600000e+04              27025.0
50%       41720.0  6.700000e+04              56211.0
75%       62580.0  1.400000e+05             100000.0
max       83439.0  1.000000e+74           45241312.0

## **Task2:** 
#### Cleaned data

In [3]:
dfP = p3.load_and_process('../data/raw/survey_results_public_2021.csv')
print(f"New data frame after cleaning data: {dfP.columns}")

New data frame after cleaning data: Index(['ResponseId', 'MainBranch', 'Employment', 'EdLevel', 'DevType',
       'LanguageHaveWorkedWith', 'DatabaseHaveWorkedWith',
       'PlatformHaveWorkedWith', 'WebframeHaveWorkedWith',
       'MiscTechHaveWorkedWith', 'ToolsTechHaveWorkedWith',
       'NEWCollabToolsHaveWorkedWith', 'ConvertedCompYearly'],
      dtype='object')


In [4]:
count = Counter()
for i, row in dfP.iterrows():
    rows = str(row['LanguageHaveWorkedWith'])
    listr = rows.split(";")
    for prog in listr:
            count[prog] += 1
dfLang = pd.DataFrame.from_dict(count, orient = 'index', columns = ['Number of people'])
print("Number of people who used the different languages: ")
dfLang.index.name = "Languages"
dfLang

Number of people who used the different languages: 


Number of people
Languages                    
C++                     20057
HTML/CSS                46259
JavaScript              53587
Objective-C              2310
PHP                     18130
Swift                    4204
Python                  39792
Assembly                 4632
C                       17329
R                        4185
Rust                     5799
TypeScript              24909
Bash/Shell              22385
SQL                     38835
C#                      22984
Java                    29162
Node.js                 27975
PowerShell               8871
Ruby                     5569
Perl                     2028
Matlab                   3846
Kotlin                   6866
Julia                    1068
Haskell                  1749
Delphi                   1731
Go                       7879
Scala                    2148
Dart                     4965
nan                      1082
VBA                      3847
Groovy                   2479
Clojure                  1552
APL                       536
LISP                     1096
F#                        804
Elixir                   1438
Erlang                    651
Crystal                   466
COBOL                     437

## **Visual representation of most and least used languages**

In [124]:
top10 = dfLang.nlargest(10, 'Number of people')
bottom10 = dfLang.nsmallest(10, 'Number of people')

In [129]:
top10

Number of people
Languages                   
JavaScript             53587
HTML/CSS               46259
Python                 39792
SQL                    38835
Java                   29162
Node.js                27975
TypeScript             24909
C#                     22984
Bash/Shell             22385
C++                    20057

In [130]:
bottom10

Number of people
Languages                  
COBOL                   437
Crystal                 466
APL                     536
Erlang                  651
F#                      804
Julia                  1068
nan                    1082
LISP                   1096
Elixir                 1438
Clojure                1552

# **Task4:** 
### Analysis